README.md
### Hadith Daily

"Hadith Daily" is an application designed to provide users with a daily dose of beautiful Hadiths (sayings and actions of the Prophet Muhammad, peace be upon him). The purpose of this app is to help users stay connected to the teachings of Islam and derive spiritual inspiration from the timeless wisdom contained in the Hadith literature.

The app allows users to register and create an account. Once registered, users will receive an email every day at 7 AM containing a randomly selected Hadith. The Hadiths are carefully curated to encompass various aspects of life, including faith, character, morality, kindness, and more. Each email will present a new Hadith along with its corresponding reference, enabling users to explore further if desired.

The goal of "Hadith Daily" is to promote a deeper understanding and appreciation of the teachings of the Prophet Muhammad, fostering a positive impact on the lives of users. By receiving these daily reminders, users can incorporate the teachings of the Hadiths into their daily routines, reflect on their meanings, and strive to embody the virtues and values they highlight.

Through the convenience of email delivery, users can seamlessly integrate these meaningful Hadiths into their morning routine, providing spiritual nourishment and guidance throughout their day. Whether it's a short yet profound saying or an inspiring anecdote from the life of the Prophet Muhammad(peace be upon him), each Hadith is carefully chosen to uplift and enlighten the hearts and minds of users.

By delivering Hadiths straight to their inbox, "Hadith Daily" aims to serve as a valuable resource for individuals seeking spiritual growth, knowledge, and a stronger connection to their faith. It is an app that encourages users to embark on a journey of self-improvement, fostering a deeper understanding of Islam's teachings and their practical application in everyday life.

Ultimately, "Hadith Daily" endeavors to create a positive impact by spreading the beautiful teachings of the Prophet Muhammad(peace be upon him) and nurturing a community of individuals who seek to align their lives with the principles of Islam.

With love and duas,

Your Brother,

Muhammad Hossain Chisty

## **Live**

![dark-mode-version](https://github.com/hossainchisty/HadithDaily/assets/62835101/e4d7c4a2-5dd1-4ee1-b1f3-c2d2b7307c07)

Visit [HadithDaily](https://hadith-daily.vercel.app/)



CHANGELOG.md
## [0.4.2](https://github.com/hossainchisty/HadithDaily/compare/v0.4.1...v0.4.2) (2023-06-12)


### Bug Fixes

* Added check to redirect dashboard if user has token on cookies ([2cc6bc7](https://github.com/hossainchisty/HadithDaily/commit/2cc6bc75d53ec1115714856b24e7be0adc09d120))



## [0.4.1](https://github.com/hossainchisty/HadithDaily/compare/v0.4.0...v0.4.1) (2023-06-12)


### Bug Fixes

* User can view dashboard if they login ([7d1a4c8](https://github.com/hossainchisty/HadithDaily/commit/7d1a4c80c3c5fc813dab4af152714e61061afa41))



# [0.4.0](https://github.com/hossainchisty/HadithDaily/compare/v0.3.0...v0.4.0) (2023-06-12)


### Features

* Added new hadiths `v2` ([6119607](https://github.com/hossainchisty/HadithDaily/commit/61196072294b41c045a62d74920a88c9d2055d9d))



# [0.3.0](https://github.com/hossainchisty/HadithDaily/compare/v0.2.0...v0.3.0) (2023-06-12)


### Features

* Added list of hadiths `v1` ([2fdf5c4](https://github.com/hossainchisty/HadithDaily/commit/2fdf5c46085bf4da86010e1a2656f9150cfae75e))



# [0.2.0](https://github.com/hossainchisty/HadithDaily/compare/v0.1.1...v0.2.0) (2023-06-12)


### Features

* Added dynamic data to dashboard ([48186f0](https://github.com/hossainchisty/HadithDaily/commit/48186f08c0a7eb4b70bfe607e8d015d6bce09fdd))





app.js
// Basic Lib Imports
require("dotenv").config();
const path = require('path');
const express = require('express');
const bodyParser = require('body-parser');
const mongoose = require('mongoose');
const cookieParser = require('cookie-parser');
const schedule = require('node-schedule');
const generateRandomHadith = require('./services/getRandomHadith');
const sendEmail = require('./services/mailer');
const User = require('./schema/userModels');


const app = express();

// Set up body parser middleware
app.use(bodyParser.urlencoded({ extended: true }));
app.use(bodyParser.json());

// Serve static files from the 'public' directory
app.use(express.static('public'));

// Set the view engine to EJS
app.set('view engine', 'ejs');
app.set('views', path.join(__dirname, 'views'));


// Set up cookie parser middleware
app.use(cookieParser());


// MongoDB connection setup
mongoose.connect(process.env.MONGODB_URL, {
  useNewUrlParser: true,
  useUnifiedTopology: true,
});

const db = mongoose.connection;

db.on('error', console.error.bind(console, 'MongoDB connection error:'));
db.once('open', () => {
  console.log('Connected to MongoDB');
});

// Schedule the task to run at 7am daily
// * * * * *  0 7 * * *
// const job = schedule.scheduleJob('* * * * * ', async () => {
//   try {
//       // Retrieve all user email addresses from the user database
//       const users = await User.find({}, 'email');

//       for (const user of users) {
//           const userEmailAddress = user.email;

//           const hadith = await generateRandomHadith();
//           const subject = 'Hadith Daily';
//           const text = `${hadith.quote}\n(Reference: ${hadith.reference})`;

//           await sendEmail(userEmailAddress, subject, text);
//       }
//   } catch (error) {
//       console.error(error);
//   }
// });



// async function sendDailyEmail() {
//   const currentDate = new Date();
//   const desiredTime = new Date(currentDate);
//   desiredTime.setUTCHours(1, 0, 0); // Set the desired time to 1:00 AM UTC time (equivalent to 7:00 AM in Bangladesh time zone)

//   const timeDiff = desiredTime - currentDate;

//   // If the desired time has passed for today, add 24 hours to the time difference to schedule for the next day
//   const timeToWait = timeDiff < 0 ? timeDiff + 24 * 60 * 60 * 1000 : timeDiff;

//   setTimeout(async () => {
//     // Retrieve all user email addresses from the user database
//     const users = await User.find({}, 'email');

//     for (const user of users) {
//       const userEmailAddress = user.email;

//       const hadith = await generateRandomHadith();
//       const subject = 'Hadith Daily';
//       const text = `${hadith.quote}\n(Reference: ${hadith.reference})`;

//       await sendEmail(userEmailAddress, subject, text);
//     }

//     // After sending the email, recursively call the function to schedule for the next day
//     sendDailyEmail();
//   }, timeToWait);
// }

// // Start the process
// sendDailyEmail();


async function sendEmailEveryMinute() {
  const users = await User.find({}, 'email');

  for (const user of users) {
    const userEmailAddress = user.email;

    const hadith = await generateRandomHadith();
    const subject = 'Hadith Daily';
    const text = `${hadith.quote}\n(Reference: ${hadith.reference})`;

    await sendEmail(userEmailAddress, subject, text);
  }

  setTimeout(sendEmailEveryMinute, 60000); // Send email every minute
}

// Start the process
sendEmailEveryMinute();

// Set the routes
app.use('/', require('./routers/index'));

// Start the server
const port = process.env.PORT || 3000;
app.listen(port, () => {
  console.log(`Server started on port ${port}`);
});



quotes.js
module.exports = [
    "Code is like humor. When you have to explain it, it's bad. - Cory House",
    "It's not a bug, it's an undocumented feature. - Anonymous",
    "Programs must be written for people to read, and only incidentally for machines to execute. - Harold Abelson",
    "The best error message is the one that never shows up. - Thomas Fuchs",
    "Software is a great combination of artistry and engineering. - Bill Gates",
    "Simplicity is the soul of efficiency. - Austin Freeman",
    "Programming is the closest thing we have to a superpower. - Drew Houston",
    "The function of good software is to make the complex appear to be simple. - Grady Booch",
    "Innovation distinguishes between a leader and a follower. - Steve Jobs",
    "The only way to do great work is to love what you do. - Steve Jobs",
  ];
  

routers/index.js
// Basic Lib Imports
const express = require('express');
const router = express.Router();
const bcrypt = require('bcryptjs');
const jwt = require('jsonwebtoken');
const Hadith = require('../schema/hadithModels');
const User = require('../schema/userModels');

const generateRandomHadith = require('../services/getRandomHadith');

// Register route
router.post('/', async (req, res) => {
    try {
        const { name, email, password } = req.body;

        // Check if the user already exists
        let user = await User.findOne({ email });
        if (user) {
            return res.render('register/register.ejs', { message: 'User already exists' });
        }

        // Hash the password
        const salt = await bcrypt.genSalt(10);
        const hashedPassword = await bcrypt.hash(password, salt);

        // Create a new user
        user = new User({
            name,
            email,
            password: hashedPassword,
        });

        // Save the user to the database
        await user.save();
        res.redirect('/registration-success');
    } catch (error) {
        console.error(error);
        res.status(500).json({ message: 'Server Error' });
    }
});

// Login route
router.post('/login', async (req, res) => {
    try {
        const { email, password } = req.body;

        // Check if the user exists
        const user = await User.findOne({ email });
        if (!user) {
            errorMessage = 'User not found';
            return res.render('login', { errorMessage });
        }

        // Check the password
        const isMatch = await bcrypt.compare(password, user.password);
        if (!isMatch) {
            errorMessage = 'Invalid credentials';
            return res.render('login', { errorMessage });
        }

        // Generate a JWT token
        const token = jwt.sign({ userId: user._id }, 'secretkey');

        // Set the token as a cookie
        res.cookie('token', token, {
            httpOnly: true,
            maxAge: 24 * 60 * 60 * 1000, // 1 day
        });

        res.redirect('/dashboard');
    } catch (error) {
        console.error(error);
        res.status(500).json({ message: 'Server Error' });
    }
});




// Protected route - render the dashboard
router.get('/dashboard', async (req, res) => {
    const token = req.cookies.token;
    if (!token) {
        return res.redirect('/login');
    }

    try {
        const decoded = jwt.verify(token, 'secretkey');
        const userId = decoded.userId;


        const hadith = await generateRandomHadith();
        const quote = hadith.quote;
        const reference = hadith.reference;

        res.render('dashboard', { quote, reference });
    } catch (error) {
        console.error(error);
        res.redirect('/login');
    }
});

// Render the registration form
router.get('/', (req, res) => {
    // Check if the user has a valid session or token
    if (req.cookies.token) {
        return res.redirect('/dashboard');

    }
    res.render('register/register.ejs', { message: '' });
});

// Render the login form
router.get('/login', (req, res) => {
    // Check if the user has a valid session or token
    if (req.cookies.token) {
        return res.redirect('/dashboard');
    }
    res.render('login', { errorMessage: '' });
});

// Render the registration success page
router.get('/registration-success', (req, res) => {
    res.render('register/registration-success.ejs');
});

// API endpoint to insert hadiths into the database
router.post('/hadiths', async (req, res) => {
    try {
        const hadiths = require('../hadiths.json'); // Load the hadiths from the JSON file

        // Insert hadiths into the database using insertMany
        const result = await Hadith.insertMany(hadiths);
        res.status(200).json({ message: 'Hadiths inserted successfully!', count: result.length });
    } catch (error) {
        console.error('Error inserting hadiths:', error);
        res.status(500).json({ message: 'Failed to insert hadiths', error: error.message });
    }
});

module.exports = router;
